flow/datatypes/numeric.go

```diff
@@ -1,111 +1,163 @@
 }
 
-func (SnowflakeNumericCompatibility) MaxScale() int16 {
-	return 37
+func getMaxPrecisionForDWH(dwh protos.DBType) int16 {
+	precision, ok := precisionMap[dwh]
+	if !ok {
+		return genericPrecision
+	}
+	return precision
 }
 
-func (s SnowflakeNumericCompatibility) DefaultPrecisionAndScale() (int16, int16) {
-	return s.MaxPrecision(), PeerDBSnowflakeScale
+func getMaxScaleForDWH(dwh protos.DBType) int16 {
+	scale, ok := maxScaleMap[dwh]
+	if !ok {
+		return genericMaxScale
+	}
+	return scale
 }
 
-type BigQueryNumericCompatibility struct{}
-
-func (BigQueryNumericCompatibility) MaxPrecision() int16 {
-	return 38
+func getDefaultPrecisionAndScaleForDWH(dwh protos.DBType) (int16, int16) {
+	defaultScale, ok := defaultScaleMap[dwh]
+	if !ok {
+		return getMaxPrecisionForDWH(dwh), genericDefaultScale
+	}
+	return getMaxPrecisionForDWH(dwh), defaultScale
 }
 
-func (BigQueryNumericCompatibility) MaxScale() int16 {
-	return 20
+func isValidPrecision(precision int16, dwh protos.DBType) bool {
+	return precision > 0 && precision <= getMaxPrecisionForDWH(dwh)
 }
 
-func (b BigQueryNumericCompatibility) DefaultPrecisionAndScale() (int16, int16) {
-	return b.MaxPrecision(), PeerDBBigQueryScale
+func isValidScale(precision int16, scale int16, dwh protos.DBType) bool {
+	return scale >= 0 &&
+		isValidPrecision(precision, dwh) &&
+		scale <= getMaxScaleForDWH(dwh)
 }
 
-type DefaultNumericCompatibility struct{}
+/*
+As far as my understanding from Postgres source code goes:
+ 1. typmod itself is a 32-bit integer.
+ 2. In the case of NUMERICs, it will be -1 or > VARHDRSZ.
+ 3. If it is -1, it means that the precision and scale are not specified and it is an "unconstrained" NUMERIC.
+ 4. If it is > VARHDRSZ, it means that the precision is specified and scale MAY be specified. Otherwise, scale defaults to 0.
+ 5. This is a "constrained" NUMERIC. Precision in this case ranges only from 1 to 1000, far less than the unconstrained limit.
+ 6. The scale in this case ranges from -1000 to 1000. Yes, it can be negative. Yes, it can be more than the precision.
 
-func (DefaultNumericCompatibility) MaxPrecision() int16 {
-	return 38
+Currently, no DWH supports the two weird cases of scales in Postgres NUMERICs. Expected is that the 0 <= scale < precision.
+In this case, we will default to the default scale and maximum precision for the DWH.
+*/
+type NumericTypmod struct {
+	constrained bool
+	precision   int16
+	scale       int16
 }
 
-func (DefaultNumericCompatibility) MaxScale() int16 {
-	return 37
-}
-
-func (DefaultNumericCompatibility) DefaultPrecisionAndScale() (int16, int16) {
-	return 38, 20
-}
+// This is to reverse what make_numeric_typmod of Postgres does:
+// logic copied from: https://github.com/postgres/postgres/blob/c4d5cb71d229095a39fda1121a75ee40e6069a2a/src/backend/utils/adt/numeric.c#L929
+// Maps most "invalid" typmods to be unconstrained (same as -1)
+func NewParsedNumericTypmod(typmod int32) *NumericTypmod {
```

Comment:
```suggestion
func NewParsedNumericTypmod(typmod int32) NumericTypmod {
```
Seems like a type best passed by value & treated as immutable

Reply:
That'd be handled by a Valid bool